src/material-experimental/mdc-chips/chip-action.ts

```diff
@@ -0,0 +1,154 @@
+
+/** @docs-private */
+@Directive({
+  selector: '[matChipAction]',
+  inputs: ['disabled', 'tabIndex'],
+  host: {
+    '[attr.tabindex]': 'disabled ? null : tabIndex',
+    '[attr.disabled]': "disabled ? '' : null",
+    '[attr.aria-disabled]': 'disabled',
+    '(click)': '_handleClick()',
+    '(keydown)': '_handleKeydown($event)',
+  },
+})
+export class MatChipAction
+  extends _MatChipActionMixinBase
+  implements AfterViewInit, OnDestroy, CanDisable, HasTabIndex, OnChanges
+{
+  private _document: Document;
+  private _foundation: MDCChipActionFoundation;
+  private _adapter: MDCChipActionAdapter = {
+    focus: () => this.focus(),
+    getAttribute: (name: string) => this._elementRef.nativeElement.getAttribute(name),
+    setAttribute: (name: string, value: string) => {
+      // MDC tries to update the tabindex directly in the DOM when navigating using the keyboard
+      // which overrides our own handling. If we detect such a case, assign it to the same property
+      // as the Angular binding in order to maintain consistency.
+      if (name === 'tabindex') {
+        this._updateTabindex(parseInt(value));
+      } else {
+        this._elementRef.nativeElement.setAttribute(name, value);
+      }
+    },
+    removeAttribute: (name: string) => {
+      if (name !== 'tabindex') {
+        this._elementRef.nativeElement.removeAttribute(name);
+      }
+    },
+    getElementID: () => this._elementRef.nativeElement.id,
+    emitEvent: <T>(eventName: string, data: T) => {
+      emitCustomEvent<T>(this._elementRef.nativeElement, this._document, eventName, data, true);
+    },
+  };
+
+  _handleClick() {
+    // Usually these events can't happen while the chip is disabled since the browser won't
+    // allow them which is what MDC seems to rely on, however the event can be faked in tests.
+    if (!this.disabled) {
+      this._foundation.handleClick();
+    }
+  }
+
+  _handleKeydown(event: KeyboardEvent) {
+    // Usually these events can't happen while the chip is disabled since the browser won't
+    // allow them which is what MDC seems to rely on, however the event can be faked in tests.
+    if (!this.disabled) {
+      this._foundation.handleKeydown(event);
+    }
+  }
+
+  protected _createFoundation(adapter: MDCChipActionAdapter): MDCChipActionFoundation {
+    return new MDCChipPrimaryActionFoundation(adapter);
+  }
+
+  constructor(
+    public _elementRef: ElementRef,
+    @Inject(DOCUMENT) _document: any,
+    private _changeDetectorRef: ChangeDetectorRef,
+  ) {
+    super();
+    this._foundation = this._createFoundation(this._adapter);
+
+    // Set the action classes here since they're somewhat long
+    // and error prone, and we need them in several places.
+    const classList = _elementRef.nativeElement.classList;
+    const actionType = this._foundation.actionType();
+    classList.add('mdc-evolution-chip__action', 'mat-mdc-chip-action');
```

Comment:
I don't 100% follow- why not set these classes in the `host` block?

Reply:
There were two reasons when I was working on this initially:
1. We still had to support ViewEngine which meant that I would've had to repeat the class bindings in two other places.
2. I wouldn't have been able to use the enum values in the `host` block.

I _think_ that both of these issues are resolved with Ivy though.